File: PERL/Cap08/c08ex06.pl
# c08ex06.pl

#!/usr/bin/perl

use strict;
use warnings;

print "PESQUISA BINARIA DE NUMEROS\n\n";

my @NUMERO;

for (my $I = 1; $I <= 10; $I++) {
  printf "Entre o %2do. numero: ", $I;
  chomp($NUMERO[$I] = <STDIN>);
}

for (my $I = 1; $I <= 9; $I++) {
   for (my $J = $I + 1; $J <= 10; $J++) {
    if ($NUMERO[$I] > $NUMERO[$J]) {
      my $X = $NUMERO[$I];
      $NUMERO[$I] = $NUMERO[$J];
      $NUMERO[$J] = $X;
    }
  }
}

my $RESP = "SIM";
while ($RESP eq "SIM") {
  print "\n";
  printf "Entre o numero a ser pesquisado: ";
  chomp(my $PESQ = <STDIN>);

  my $COMECO = 1;
  my $FINAL = 10;
  my $ACHA = 0;
  my $MEIO;
  while ($COMECO <= $FINAL && $ACHA == 0) {
    $MEIO = int(($COMECO + $FINAL) / 2);
    if ($PESQ == $NUMERO[$MEIO]) {
      $ACHA = 1;
    } else {
      if ($PESQ < $NUMERO[$MEIO]) {
          $FINAL = $MEIO - 1;
      } else {
          $COMECO = $MEIO + 1;
      }
    }
  }

  if ($ACHA == 1) {
    print "\n";
    print $PESQ . " foi localizado na posicao " . $MEIO . "\n";
  } else {
    print "\n";
    print $PESQ . " nao foi localizado\n";
  }

  print "\n";
  inicio:
    printf "Deseja continuar? (SIM/NAO): ";
    chomp(my $RESPOSTA = <STDIN>);
    $RESP = uc($RESPOSTA);
    if ($RESP eq "SIM" || $RESP eq "NAO") {
      goto fim;
    } else {
      print "Informe apenas SIM ou NAO.\n";
      print "\n";
    }
  goto inicio;
  fim:
}

print "\n";
printf "Tecle <Enter> para encerrar... ";
<STDIN>;

File: RUBY/Cap08/c08ex06.rb
# c08ex06.rb

#!/usr/bin/env ruby

numero = []

puts "PESQUISA BINARIA DE numeroS\n\n"

for i in 0..9
  printf "Entre o %2do. numero: ", (i + 1)
  numero[i] = gets.chomp
end

for i in 0..8
  for j in (i + 1)..9
    if (numero[i] > numero[j])
      x = numero[i]
      numero[i] = numero[j]
      numero[j] = x
    end
  end
end

resp = "SIM"
while (resp == "SIM") do
  print "\n"
  printf "Entre o numero a ser pesquisado: "
  pesq = gets.chomp

  comeco = 0
  final = 9
  acha = false
  while (comeco <= final) && (acha == false) do
    meio = ((comeco + final) / 2).to_i
    if (pesq == numero[meio])
      acha = true
    elsif (pesq < numero[meio])
      final = meio - 1
    else
      comeco = meio + 1
    end
  end

  if (acha == true)
    print "\n"
    printf "%s foi localizado na posicao %d\n", pesq, (meio + 1)
  else
    print "\n"
    printf "%s nao foi localizado\n", pesq
  end

  print "\n"
  loop do
    printf "Deseja continuar? (SIM/NAO): "
    resposta = gets.chomp.upcase
    if (resposta == "SIM") || (resposta == "NAO")
      resp = resposta
      break
    else
      puts "Informe apenas SIM ou NAO."
      print "\n"
    end
  end
end

print "\n"
printf "Tecle <Enter> para encerrar... "
gets
